### DISCUSSION TITLE
How to build native image with swing?

### DISCUSSION THREAD

**Discussion by lcy-1024** [quarkusio/quarkus#42473] [Q&A] [answered]
run command go get "native-image-agent-final-config" files,but get "java.awt.HeadlessException"

main code:
`java.awt.EventQueue.invokeLater(() -> new NewJFrame().setVisible(true));`

build:
`mvnw verify -DskipITs=false -Dquarkus.test.integration-test-profile=test-with-native-agent`

auto generate podman command:
```
podman run --name quarkus-integration-test-EHvXd -i --rm -p 8081:8081 -p 8444:8444 --entrypoint java -v /d/work/project/xxx/ta
rget:/project:z --env QUARKUS_LOG_CATEGORY__IO_QUARKUS__LEVEL=INFO --env QUARKUS_HTTP_PORT=8081 --env QUARKUS_HTTP_SSL_PORT=8444 --env TEST_URL=http://localhost:8081 --env QUARKUS_PROFILE=test-with-native-agent --env QUARKUS_TES
T_INTEGRATION_TEST_PROFILE=test-with-native-agent quay.io/quarkus/ubi-quarkus-mandrel-builder-image:jdk-21 -agentlib:native-image-agent=access-filter-file=quarkus-access-filter.json,caller-filter-file=quarkus-caller-filter.json,config-output-dir=native-image-agent-base-config, -jar xxx.jar
```

build failed,the logs is "quarkus.log":
> Exception in thread "AWT-EventQueue-0" java.awt.HeadlessException: 
> No X11 DISPLAY variable was set,
> or no headful library support was found,
> but this program performed an operation which requires it,
> 
> 	at java.desktop/java.awt.GraphicsEnvironment.checkHeadless(GraphicsEnvironment.java:164)
> 	at java.desktop/java.awt.Window.<init>(Window.java:553)
> 	at java.desktop/java.awt.Frame.<init>(Frame.java:428)
> 	at java.desktop/java.awt.Frame.<init>(Frame.java:393)
> 	at java.desktop/javax.swing.JFrame.<init>(JFrame.java:180)
> 	at com.lcy.NewJFrame.<init>(NewJFrame.java:15)
> 	at com.lcy.NewJFrame.lambda$main$0(NewJFrame.java:91)
> 	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
> 	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:773)
> 	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:720)
> 	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:714)
> 	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
> 	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
> 	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
> 	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
> 	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
> 	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
> 	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
> 	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
> 	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

actually it generate json files
**_native-image-agent-final-config_**

and then I go on build:
`mvnw verify -Dnative -Dquarkus.native.agent-configuration-apply`

auto generate build cmd is:
```
D:\work\java\graalvm-jdk-21.0.3+7.1\bin\native-image.cmd -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Dlog
ging.initial-configurator.min-level=500 -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.
allocator.maxOrder=3 -J-Duser.language=zh -J-Duser.country=CN -J-Dfile.encoding=UTF-8 --features=io.quarkus.runner.Feature,io.quarkus.runtime.graal.DisableLoggingFeature -J--add-exports=java.security.jgss/sun.security.krb5=ALL-U
NNAMED -J--add-exports=java.security.jgss/sun.security.jgss=ALL-UNNAMED -J--add-opens=java.base/java.text=ALL-UNNAMED -J--add-opens=java.base/java.io=ALL-UNNAMED -J--add-opens=java.base/java.lang.invoke=ALL-UNNAMED -J--add-opens
=java.base/java.util=ALL-UNNAMED -H:+UnlockExperimentalVMOptions -H:BuildOutputJSONFile=xxx-1.0-SNAPSHOT-runner-build-output-stats.json -H:-UnlockExperimentalVMOptions --strict-image-heap -H:+UnlockExperimentalVMOptions -H
:+AllowFoldMethods -H:-UnlockExperimentalVMOptions -J-Djava.awt.headless=true --no-fallback --link-at-build-time -H:+UnlockExperimentalVMOptions -H:+ReportExceptionStackTraces -H:-UnlockExperimentalVMOptions -H:-AddAllCharsets -
-enable-url-protocols=http --enable-monitoring=heapdump -H:+UnlockExperimentalVMOptions -H:-UseServiceLoaderFeature -H:-UnlockExperimentalVMOptions -J--add-exports=org.graalvm.nativeimage/org.graalvm.nativeimage.impl=ALL-UNNAMED
 -H:ConfigurationFileDirectories=native-image-agent-config --exclude-config io\.netty\.netty-codec /META-INF/native-image/io\.netty/netty-codec/generated/handlers/reflect-config\.json --exclude-config io\.netty\.netty-handler /META-INF/native-image/io\.netty/netty-handler/generated/handlers/reflect-config\.json xxx-1.0-SNAPSHOT-runner -jar xxx-1.0-SNAPSHOT-runner.jar
```

failed output:

> Error: An object of type 'sun.nio.ch.FileChannelImpl' was found in the image heap. This type, however, is marked for initialization at image run time for the following reason: classes are initialized at run time by default.     
> This is not allowed for correctness reasons: All objects that are stored in the image heap must be initialized at build time.
> 
> You now have two options to resolve this:
> 
> 1) If it is intended that objects of type 'sun.nio.ch.FileChannelImpl' are persisted in the image heap, add
> 
>     '--initialize-at-build-time=sun.nio.ch.FileChannelImpl'
> 
> to the native-image arguments. Note that initializing new types can store additional objects to the heap. It is advised to check the static fields of 'sun.nio.ch.FileChannelImpl' to see if they are safe for build-time initialization,  and that they do not contain any sensitive data that should not become part of the image.
> 
> 2) If these objects should not be stored in the image heap, you can use
> 
>     '--trace-object-instantiation=sun.nio.ch.FileChannelImpl'
> 
> to find classes that instantiate these objects. Once you found such a class, you can mark it explicitly for run time initialization with
> 
>     '--initialize-at-run-time=<culprit>'
> 
> to prevent the instantiation of the object.
> 
> If you are seeing this message after enabling '--strict-image-heap', this means that some objects ended up in the image heap without their type being marked with --initialize-at-build-time.
> To fix this, include '--initialize-at-build-time=sun.nio.ch.FileChannelImpl' in your configuration. If the classes do not originate from your code, it is advised to update all library or framework dependencies to the latest version before addressing this error.
> Please address this problem to be prepared for future releases of GraalVM.
> 
> The following detailed trace displays from which field in the code the object was reached.
> Trace: Object was reached by
>   reading field sun.font.TrueTypeFont$TTDisposerRecord.channel of constant
>     sun.font.TrueTypeFont$TTDisposerRecord@3dcfea04: sun.font.TrueTypeFont$TTDisposerRecord@3dcfea04
>   reading field java.util.Hashtable$Entry.value of constant
>     java.util.Hashtable$Entry@5a681d39: java.lang.ref.WeakReference@17ddf7b=sun.font.TrueTypeFont$TTDisposerRecord@3dcfe...
>   indexing into array java.util.Hashtable$Entry[]@1512d1c3: [Ljava.util.Hashtable$Entry;@1512d1c3
>   reading field java.util.Hashtable.table of constant
>     java.util.Hashtable@1658d185: {java.lang.ref.WeakReference@3168b7ff=sun.font.TrueTypeFont$TTDisposerRecord@961...
>   scanning root java.util.Hashtable@1658d185: {java.lang.ref.WeakReference@3168b7ff=sun.font.TrueTypeFont$TTDisposerRecord@961... embedded in
>     sun.java2d.Disposer.run(Disposer.java:147)
>   parsing method sun.java2d.Disposer.run(Disposer.java:145) reachable via the parsing context
>     at java.util.OptionalInt.ifPresentOrElse(OptionalInt.java:185)
>     at root method.(Unknown Source)
> ......

[a comment by an automated account is not shown]

**Comment by Karm** [comment]
Hello @lcy-1024, there is no plan to support desktop integration with Quarkus Native at the moment, left alone whole Swing toolkit. All our image processing, graphics etc. endeavors revolve around **headless** usage, i.e. server side rendering, documents processing, documents conversions etc.

Could you share what is the use case you would like to solve? What is it you need the app to do for the user?

**Reply by lcy-1024** [reply]
Hi Karm, a native image can't be decompile,it's more safer, and don't need run with JRE so that it's more convenient and faster.Java have desktop app ability(swing or javafx),why not try to make a java native image desktop app:)

**Reply by Karm** [reply, marked as answer]
@lcy-1024 
Decompilation:
Native-image is easily decompiled, for convenience you can use Ghirda or paid IDA Pro (no experience with that one); Frida scripts also work. We do not attempt any assembly obfuscation or anti-analysis or anti-sandbox tricks. If you hide a proprietary knowledge in the executable it is going to be discovered. The fact that you don't get Java bytecode back does not help to hide much.
Please, do not assume a native-image compilation of a Java program protects your secrets in any way.

Faster:
Yes, the app would have a faster satrtup-time, a rich GUI app would be very specific in how much you have to init at runtime anyway, so the benefit depends... You'd have to measure that.

No JRE:
Yes, sharing a single executable and a handful of .so (.dll/.lib) files is definitely convenient. 

You might want to check https://gluonhq.com/products/javafx/

Let me close this question as it is not Quarkus related.
